src/ngx_http_lua_upstream_module.c

```diff
@@ -98,10 +84,854 @@ ngx_http_lua_upstream_create_module(lua_State * L)
+            lua_pushnil(L);
+            lua_pushliteral(L, "the peer is exist\n");
+            return 2;
+        }
+
+        n = peers->number -1 ;
+        n += peers->next != NULL ? peers->next->number : 0;
+        old_size = n*sizeof(ngx_http_upstream_rr_peer_t) 
+			+ sizeof(ngx_http_upstream_rr_peers_t)*(peers->next != NULL ? 2:1);
+        new_size = sizeof(ngx_http_upstream_rr_peer_t) + old_size;
+    
+        peers  = ngx_prealloc(ngx_cycle->pool, uscf->peer.data, old_size, new_size );
+        if (NULL == peers ) {
+            lua_pushnil(L);
+            lua_pushliteral(L, "peers pcalloc fail\n");
+            return 2;
+        }
+
+
+        peer.weight = us->weight; 
+        peer.effective_weight = us->weight;
+        peer.current_weight= 0;
+        peer.max_fails = us->max_fails;
+        peer.fail_timeout = us->fail_timeout;
+        peer.sockaddr = us->addrs->sockaddr;
+        peer.socklen = us->addrs->socklen;
+        peer.name = us->addrs->name;
+        peer.down = us->down;
+        peer.fails = 0;
+
+        peer.server = us->name;
+
+        peers->peer[peers->number++] = peer;
+        peers->total_weight += peer.weight;
+        peers->single = (peers->number == 1);
+        peers->weighted = (peers->total_weight != peers->number);
+
+        uscf->peer.data = peers;
+    } else {
+
+        backup = peers->next;
+        if ( 1 == ngx_http_lua_upstream_exist_peer(backup,u) ) {
+            lua_pushnil(L);
+            lua_pushliteral(L, "the backup peer is exist\n");
+            return 2;
+        }
+
+        n = backup != NULL ? (backup->number - 1) : 0 ;
+
+        old_size = n*sizeof(ngx_http_upstream_rr_peer_t) 
+			       + sizeof(ngx_http_upstream_rr_peers_t);
+        new_size = sizeof(ngx_http_upstream_rr_peer_t) + old_size;
+    
+        backup  = ngx_prealloc(ngx_cycle->pool, peers->next, old_size, new_size );
+        if (NULL == backup ) {
+            lua_pushnil(L);
+            lua_pushliteral(L, "backup pcalloc fail\n");
+            return 2;
+        }
+
+        peers->single = 0;
+        backup->single = 0;
+        if ( n == 0 ) {
+           backup->number = 0;
+           backup->total_weight = 0;
+        }
+        peer.weight = us->weight; 
+        peer.effective_weight = us->weight;
+        peer.current_weight= 0;
+        peer.max_fails = us->max_fails;
+        peer.fail_timeout = us->fail_timeout;
+        peer.server = us->name;
+
+        peer.sockaddr = us->addrs->sockaddr;
+        peer.socklen = us->addrs->socklen;
+        peer.name = us->addrs->name;
+        peer.down = us->down;
+        peer.fails = 0;
+   
+
+        backup->peer[backup->number++] = peer;
+        backup->total_weight += peer.weight;
+        backup->single = (backup->number == 1);
+        backup->weighted = (backup->total_weight != backup->number);
+
+        peers->next = backup;
+}
+
     return 1;
+
+}
+
+
+
+/*
+ * The function copy from tengine-2.1.0 core/ngx_palloc.c 
```

Comment:
@wangfakang I am afraid even for Tengine's version of ngx_prealloc, there is a risk of memory leaks due to potential fragmentation in the pool in the long run. And that's probably why nginx does not provide such things in the first place (and the closest thing is `ngx_array_t` which is specialized data structure though). I have no better suggestions on this yet and I wonder if we shouldn't do such things in the first place.